### PR TITLE
Feat/tracking feats

### PR DESCRIPTION
**Pull request by fight4dream**
feat(Tracking): add new velocity trackers
feat(Tracking): extract as IProcessable for two components

### CODE REVIEW

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        /// Determines whether to extract the local property or the world property.
+        /// </summary>
+        [Serialized]
+        [field: DocumentedByXml]
+        public bool UseLocal { get; set; }
+
+        /// <inheritdoc />
+        protected override Vector3 DoGetAngularVelocity()
+        {
+            return UseLocal ? transform.localRotation * AngularVelocity : AngularVelocity;
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:36]: isn't this check pointless? why even have the UseLocal on this type?

File: Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        [Serialized, Cleared]
+        [field: DocumentedByXml]
+        public Rigidbody Source { get; set; }
+
+        /// <summary>
+        /// The velocity to return.
+        /// </summary>
+        [Serialized]
+        [field: DocumentedByXml]
+        public Vector3 Velocity { get; set; }
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs:25]: why can you specify a velocity (and angular velocity)? these don't seem to be used anywhere, only Source is used.

**Review by thestonefox — CHANGES_REQUESTED** [review]

File: Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs [hunk 3]
```diff
@@ -1,189 +1,55 @@
 ﻿namespace Zinnia.Tracking.Velocity
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs:1]: Why is this being extracted? This class doesn't use an Update method to process anything it simply calls Apply. So if anything the wrapper should be the other way round no?

This should stay as is, and if you want something that is iProcessable then you just have it call the Apply method in this class on every moment?

**Review by thestonefox — CHANGES_REQUESTED** [review]

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿namespace Zinnia.Tracking.Velocity
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:1]: what's the use case for this class?

**Review by thestonefox — CHANGES_REQUESTED** [review]

File: Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs (hunk 2 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs:25]: Ah i forgot to remove those properties after dup from constant velocity tracker
It should only have the Source property

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 4 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:1]: My move in place uses velocity trackers as speed input. To debug an issue, i use this class to isolate the cause and find out average velocity tracker is part of the problem (frame rate dependant)

Other than debug purpose, this class is for “velocity tracker” whereas “Constant Force” component is for rigidbody. For example, for simulator/non vr mode, I had to simulate a grab and throw. Without this class, we had to key an animation for the interactor to release the object at certain angle and speed. It is a hard way. With this class, we create a constant tracker parented to playAreaAlias, with UseLocal=true, Velocity=(0,0.5,0.5), the object will be thrown consistently to direction desired and speed desired.

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 1 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:36]: It is analogous to the Constant Force component’s world space force or relative force. 
Here we specify whether the velocity is world space or relative to self 
Also see above simulated throw use case.

File: Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs (hunk 3 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs:1]: Notice originally when call Apply() once, it starts a coroutine to “process” the deceleration every frame (coroutine works post Update() but before LateUpdate()). So it is NOT equivalent to call Apply() in the moment we want to.

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 4 above)
Comment (thestonefox) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:1]: ok. this class needs a complimenting test to go along with it

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 1 above)
Comment (thestonefox) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:36]: sorry, I read the line wrong, I didn't see the `transform.localRotation *`

this also needs a test for the class

File: Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+﻿namespace Zinnia.Tracking.Velocity
+{
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs:2]: class needs a test

**Review by thestonefox — CHANGES_REQUESTED** [review]

File: Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs (hunk 3 above)
Comment (thestonefox) [reply Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs:1]: right I see, so its the Deceleration is the thing that is being moved into the Process, I understand now

Comment (thestonefox) [inline Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs:1]: Do the tests still pass for these updates?

Comment (fight4dream) [reply Runtime/Tracking/Velocity/ArtificialVelocityApplier.cs:1]: image
yes they are.

File: Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs (hunk 5 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs:2]: image
will force push test class when all done

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 1 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:36]: image

File: Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs (hunk 2 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/RigidbodyVelocityTracker.cs]: removed

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 4 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:1]: done

**Comment by fight4dream** [comment]
image

**Comment by fight4dream** [comment]
@thestonefox is this what you wanted?

**Comment by thestonefox** [comment]
> @thestonefox is this what you wanted?

No, I mean just delete that last commit, I'll do a commit that fixes that stuff

**Comment by fight4dream** [comment]
done~

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        /// <inheritdoc />
+        protected override Vector3 DoGetAngularVelocity()
+        {
+            return UseLocal ? transform.localRotation * AngularVelocity : AngularVelocity;
+        }
+
+        /// <inheritdoc />
+        protected override Vector3 DoGetVelocity()
+        {
+            return UseLocal ? transform.localRotation * Velocity : Velocity;
```
Comment (thestonefox) [inline Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:42]: why is this multiplying by local rotation for velocity?:

**Review by thestonefox — CHANGES_REQUESTED** [review]

File: Runtime/Tracking/Velocity/ConstantVelocityTracker.cs (hunk 6 above)
Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:42]: https://github.com/ExtendRealityLtd/Zinnia.Unity/pull/520#discussion_r530399569

Comment (fight4dream) [reply Runtime/Tracking/Velocity/ConstantVelocityTracker.cs:42]: https://github.com/ExtendRealityLtd/Zinnia.Unity/pull/520#discussion_r530400950

**Pull request merged**